File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Source/Tva.Phasors/ChannelFrameParsingStateBase.vb
'*******************************************************************************************************
'  ChannelFrameParsingStateBase.vb - Channel data frame parsing state base class
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: James R Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  01/14/2005 - James R Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

' This class represents the protocol independent common implementation the parsing state used by any frame of data that can be sent or received from a PMU.
<CLSCompliant(False)> _
Public MustInherit Class ChannelFrameParsingStateBase(Of T As IChannelCell)

    Inherits ChannelParsingStateBase
    Implements IChannelFrameParsingState(Of T)

    Private m_cells As IChannelCellCollection(Of T)
    Private m_cellCount As Integer
    Private m_parsedBinaryLength As UInt16
    Private m_createNewCellFunction As IChannelFrameParsingState(Of T).CreateNewCellFunctionSignature

    Public Sub New(ByVal cells As IChannelCellCollection(Of T), ByVal parsedBinaryLength As UInt16, ByVal createNewCellFunction As IChannelFrameParsingState(Of T).CreateNewCellFunctionSignature)

        m_cells = cells
        m_parsedBinaryLength = parsedBinaryLength
        m_createNewCellFunction = createNewCellFunction

    End Sub

    Public ReadOnly Property CreateNewCellFunction() As IChannelFrameParsingState(Of T).CreateNewCellFunctionSignature Implements IChannelFrameParsingState(Of T).CreateNewCellFunction
        Get
            Return m_createNewCellFunction
        End Get
    End Property

    Public ReadOnly Property Cells() As IChannelCellCollection(Of T) Implements IChannelFrameParsingState(Of T).Cells
        Get
            Return m_cells
        End Get
    End Property

    Public Property CellCount() As Integer Implements IChannelFrameParsingState(Of T).CellCount
        Get
            Return m_cellCount
        End Get
        Set(ByVal value As Integer)
            m_cellCount = value
        End Set
    End Property

    Public Property ParsedBinaryLength() As UInt16 Implements IChannelFrameParsingState(Of T).ParsedBinaryLength
        Get
            Return m_parsedBinaryLength
        End Get
        Set(ByVal value As UInt16)
            m_parsedBinaryLength = value
        End Set
    End Property

End Class

File: Source/OldCode/Archive/2005 TVA Code Library/Version/4.0/Source/TVA.PhasorProtocols/ChannelFrameParsingStateBase.cs
//*******************************************************************************************************
//  ChannelFrameParsingStateBase.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R. Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  01/14/2005 - James R. Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;

namespace TVA.PhasorProtocols
{
    /// <summary>
    /// Represents the protocol independent common implementation of the parsing state used by any <see cref="IChannelFrame"/>.
    /// </summary>
    /// <typeparam name="T"><see cref="Type"/> of <see cref="IChannelCell"/> referenced by <see cref="IChannelFrame"/> being parsed.</typeparam>
    public abstract class ChannelFrameParsingStateBase<T> : ChannelParsingStateBase, IChannelFrameParsingState<T> where T : IChannelCell
    {
        #region [ Members ]

        // Fields
        private int m_cellCount;
        private CreateNewCellFunction<T> m_createNewCellFunction;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="ChannelFrameParsingStateBase{T}"/> from specified parameters.
        /// </summary>
        /// <param name="parsedBinaryLength">Binary length of the <see cref="IChannelFrame"/> being parsed.</param>
        /// <param name="createNewCellFunction">Reference to delegate to create new <see cref="IChannelCell"/> instances.</param>
        protected ChannelFrameParsingStateBase(int parsedBinaryLength, CreateNewCellFunction<T> createNewCellFunction)
        {
            base.ParsedBinaryLength = parsedBinaryLength;
            m_createNewCellFunction = createNewCellFunction;
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets reference to delegate used to create a new <see cref="IChannelCell"/> object.
        /// </summary>
        public virtual CreateNewCellFunction<T> CreateNewCell
        {
            get
            {
                return m_createNewCellFunction;
            }
        }

        /// <summary>
        /// Gets or sets number of cells that are expected in associated <see cref="IChannelFrame"/> being parsed.
        /// </summary>
        public virtual int CellCount
        {
            get
            {
                return m_cellCount;
            }
            set
            {
                m_cellCount = value;
            }
        }

        #endregion
    }
}